File: src/domain/caddyfile.ts
import type { HostPort, ProxyDomain } from "../schema/ids.js"

// ---------------------------------------------------------------------------
// domain/caddyfile — pure Caddyfile route codec + port allocation.
//
// Semantics extracted verbatim from the original ProxyService:
//   - domain-block regex                   /^([a-z0-9][a-z0-9.\-]*)\s*\{/
//   - reverse_proxy host.docker.internal   /reverse_proxy\s+host\.docker\.internal:(\d+)/
//   - removeRoute: skip block lines, collapse blank runs (\n{3,} → \n\n),
//     trim, then a single trailing newline.
//   - nextPort: BASE_PORT = 5173; lowest free port above the base (hole-filling).
//
// No services, no IO, no Effect — brands imported from src/schema/ids.ts.
// ---------------------------------------------------------------------------

export interface Route {
  readonly domain: ProxyDomain
  readonly port: HostPort
}

export const BASE_PORT = 5173

const DOMAIN_RE = /^([a-z0-9][a-z0-9.\-]*)\s*\{/
const PROXY_RE = /reverse_proxy\s+host\.docker\.internal:(\d+)/

export const parseRoutes = (content: string): ReadonlyArray<Route> => {
  if (content.trim().length === 0) return []
  const routes: Route[] = []
  const lines = content.split("\n")
  let currentDomain: string | null = null
  for (const line of lines) {
    const domainMatch = line.match(DOMAIN_RE)
    if (domainMatch) currentDomain = domainMatch[1]!
    const proxyMatch = line.match(PROXY_RE)
    if (proxyMatch && currentDomain) {
      routes.push({
        domain: currentDomain as ProxyDomain,
        port: parseInt(proxyMatch[1]!, 10) as HostPort,
      })
      currentDomain = null
    }
  }
  return routes
}

export const addRoute = (content: string, route: Route): string => {
  const block = `\n${route.domain} {\n    reverse_proxy host.docker.internal:${route.port}\n}\n`
  return content + block
}

export const removeRoute = (content: string, domain: string): string => {
  const lines = content.split("\n")
  const result: string[] = []
  let skip = false
  for (const line of lines) {
    if (line.startsWith(`${domain} {`)) {
      skip = true
      continue
    }
    if (skip && line.trim() === "}") {
      skip = false
      continue
    }
    if (!skip) result.push(line)
  }
  return result.join("\n").replace(/\n{3,}/g, "\n\n").trim() + "\n"
}

export const nextPort = (
  routes: ReadonlyArray<Route>,
  base: number = BASE_PORT,
): HostPort => {
  const used = new Set<number>(routes.map((r) => r.port))
  let port = base + 1
  while (used.has(port)) port++
  return port as HostPort
}

File: src/domain/caddyfile.rs
use regex::Regex;
use std::sync::LazyLock;

// Pure Caddyfile route codec + port allocation. Semantics ported verbatim:
//   - domain-block regex                   ^([a-z0-9][a-z0-9.\-]*)\s*\{
//   - reverse_proxy host.docker.internal   reverse_proxy\s+host\.docker\.internal:(\d+)
//   - remove_route: skip block lines, collapse blank runs (\n{3,} → \n\n),
//     trim, then a single trailing newline.
//   - next_port: lowest free port above the base (hole-filling).

#[derive(Debug, Clone)]
pub struct Route {
    pub domain: String,
    pub port: u16,
}

pub const BASE_PORT: u16 = 5173;

static DOMAIN_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^([a-z0-9][a-z0-9.\-]*)\s*\{").unwrap());
static PROXY_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"reverse_proxy\s+host\.docker\.internal:(\d+)").unwrap());
static BLANK_RUN_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\n{3,}").unwrap());

pub fn parse_routes(content: &str) -> Vec<Route> {
    if content.trim().is_empty() {
        return vec![];
    }
    let mut routes = Vec::new();
    let mut current_domain: Option<String> = None;
    for line in content.split('\n') {
        if let Some(m) = DOMAIN_RE.captures(line) {
            current_domain = Some(m[1].to_string());
        }
        if let Some(m) = PROXY_RE.captures(line) {
            if let Some(domain) = current_domain.take() {
                if let Ok(port) = m[1].parse::<u16>() {
                    routes.push(Route { domain, port });
                }
            }
        }
    }
    routes
}

pub fn add_route(content: &str, domain: &str, port: u16) -> String {
    format!("{content}\n{domain} {{\n    reverse_proxy host.docker.internal:{port}\n}}\n")
}

pub fn remove_route(content: &str, domain: &str) -> String {
    let mut result: Vec<&str> = Vec::new();
    let mut skip = false;
    let block_start = format!("{domain} {{");
    for line in content.split('\n') {
        if line.starts_with(&block_start) {
            skip = true;
            continue;
        }
        if skip && line.trim() == "}" {
            skip = false;
            continue;
        }
        if !skip {
            result.push(line);
        }
    }
    let joined = result.join("\n");
    let collapsed = BLANK_RUN_RE.replace_all(&joined, "\n\n");
    format!("{}\n", collapsed.trim())
}

pub fn next_port(routes: &[Route], base: u16) -> u16 {
    let used: std::collections::HashSet<u16> = routes.iter().map(|r| r.port).collect();
    let mut port = base + 1;
    while used.contains(&port) {
        port += 1;
    }
    port
}
